### DISCUSSION TITLE
ViewBadge - inconsistencies and more flexibility?

### DISCUSSION THREAD

**Discussion by bryceitoc9** [microsoft/vscode-discussions#955] [Extension Development QnA] [answered]
Hiya! This is half a bug report and half a feature request: I was playing around with the `ViewBadge` API a little while ago and found that there's some inconsistencies between using it on a `TreeView` and a `WebviewView`! The main inconsistencies I've seen have been:
* `WebviewView`s won't render until the provider has been activated, so if you want to send a message via it, the WebviewView will have to be opened first.
* while both type the value as a number, the `WebviewView` lets you actually use strings, negative numbers, etc. See, for instance:
![image](https://github.com/microsoft/vscode-discussions/assets/29374703/3cdf7681-f896-4a83-af39-e2a401e1ff2e)
![image](https://github.com/microsoft/vscode-discussions/assets/29374703/9db99a3f-e675-453f-a5df-c9b4eb894d8f)
  * (note: this specifically _does not_ work on `ViewBadge`s rendered by `TreeView`s, only `WebviewViews`)

Our team was looking for a way to pop a blank badge on _first load only_ since having a `1` didn't make sense. We would've loved a formal way to show a blank badge without a user interacting with the `WebviewView` that the notification was tied to; if this would've worked without interacting with the `WebviewView` first, we probably would've used the string hack.

The method we ended up using was a `TreeView` in the view container, with a `"when": false`, which let us never see the view, and also instantly throw up the `ViewBadge` (instead of waiting for interaction on a `WebviewView`).

Ultimately, I think the most pressing thing here is just to try and smooth out the wrinkles between the views. As for feature requests, I'd love it if `WebviewView`s can show the badge ahead of time, and I'd also love it if we could have a blank badge: perhaps show a blank if `value === 0`, and make an `undefined` `ViewBadge` remove the badge instead of both removing? FWIW I think allowing full strings is awful--this was mainly a proof-of-concept!

**Comment by tamuratak** [comment, marked as answer]
>  We would've loved a formal way to show a blank badge without a user interacting with the WebviewView

Tracked at https://github.com/microsoft/vscode/issues/146330.

A workaround is calling `vscode.commands.executeCommand("your_webviewview_id.focus")`, which opens the WebView view. The command is automatically registered by VS Code. You can see them in the Command Pallete or the keyboard shortcut panel. See https://github.com/microsoft/vscode/issues/106085#issuecomment-692464973
